src/ninja.cc

```diff
@@ -148,9 +149,6 @@ struct NinjaMain : public BuildLogUser {
   void DumpMetrics();
 
   virtual bool IsPathDead(StringPiece s) const {
-    Node* n = state_.LookupNode(s);
-    if (!n || !n->in_edge())
-      return false;
```

Comment:
I think this is a bug, right?

Reply:
Can you explain a little bit? If it was a bug, it should be caught by a test.

Reply:
Well, if it had been covered, removing it would have made some test fail... Which is why I am asking the question: if there was a test, I could at least have understood the purpose of this.

If we check if s is a dead path, and we don't find it, or we find it, but it in not in edge, then why would it not be considered as dead? what is the logic? What is the corner case to which it applies? 
I could not find any reason for this code... and it is exactly the definition of "dead" I used in the Cleaner::CleanDead() implementation.

Reply:
Yeah makes sense.

Re-adding those lines won't make any of your test fail though, that's what I meant by "caught by a test".

Reply:
Actually, the right answer is that I should add a test for removing these 3 lines.
I don't have the time right now... I'll do it as soon as I can (unless you really disagree with removing these 3 lines).

Basically: if the code was right, if a file used to be built, but has both been removed from the manifest and from the build dir, but still has an entry in the .ninja_log, the command:
`    ninja -t recompact
should remove the entry from the log... With these 3 lines, it does not... Which means it leaves useless stuff in the .ninja_log which goes against the idea of speed.

Also: removing these 3 lines is kind of part of the feature here: the "recompact" tool considers "dead path" as candidate to be removed from the .ninja_log (even so, with these 3 lines, the answer IsPathDead answer is always false...)... but the "clean" tool does not remove the artifacts that are no more in the manifest... so the tool "cleandead" actually remove those, thus allowing the "recompact" tool (without those 3 lines) to do its job (remove entries that are now useless because duplicated or referencing dead paths).

Ideally, if I find some time I'd really like to have the cleandead tool also perform the recompact.

In any case: let me know how you would like to proceed.

Reply:
Thanks for the explanation. I still don't understand it 100%, that's why a test would be good.

I would also merge this PR as it is without the removal of those 3 lines and handle the change in a new PR.

Reply:
> These lines were added in the following commit, but the change seems unrelated... That's why I think it was added by mistake.

Actually they weren't added but moved from:

```cpp
    return (!n || !n->in_edge()) && disk_interface_.Stat(s.AsString()) == 0;
```

But I think it's a logic error. The equivalent code would be:

```cpp
    if (n && n->in_edge())
      return false;
```

Do you agree? If so, I would fix this in a commit on master.

@nico Can you also have a look? This was the change: https://github.com/ninja-build/ninja/commit/3beebde51a2089ecb01820f1428efe0263deaeea#diff-78294872cbf9d32f4f972288561fa718R146

Reply:
I've fixed the conditional in https://github.com/ninja-build/ninja/commit/1f02bc192a31ec2daac813c11d98dd939848d40f. When you're rebasing, just keep the three lines.